enhancement(aws_s3 sink): Add ability to configure request errors to retry

jchap-pnnl opened the pull request:
## Summary
Adds a new field, a RetryStrategy enum called retry_strategy, to S3SinkConfig that allows users to specify types of response errors to retry. Users can specify specific status codes of error responses of failed requests they want to be automatically retried, or they can specify all failed requests to be automatically retried.

## Change Type
- [ ] Bug fix
- [x] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [ ] Yes
- [x] No

## How did you test this PR?
1. Ran existing integration tests and verified they passed. 
2. Manually tested to verify the new functionality. Generated 403 errors by changing the `access_key_id` to an invalid one. Verified the failed authentication service was retried when the configuration required it to and didn't when the configuration didn't require it to. Ran tests with different values for `retry_strategy`, `type`, and `status_codes`:
    - retry_strategy: omitted |
        - type: all | custom
            - status_codes: [] | [403] | [404] | [403, 404]

Used this configuration: 
```
# vector.yaml
sources:
  generate_syslog:
    type: "demo_logs"
    format: "syslog"
    count: 10

transforms:
  remap_syslog:
    inputs:
      - "generate_syslog"
    type: "remap"
    source: |
      structured = parse_syslog!(.message)
      . = merge(., structured)

sinks:
  s3:
    type: aws_s3
    encoding:
      codec: "json"
    region: <actual region>
    inputs:
      - remap_syslog
    bucket: <bucket name>
    auth:
      access_key_id: <key id value>
      secret_access_key: <key value>

    retry_strategy: 
      type: all
      # type: custom
      #   status_codes: [403]
```
3. Considered adding a new test to src/sinks/aws_s3/integration_tests.rs, but didn't see a good option for detecting service retries. Considered a test using a timer that could measure how long a successful request takes and determine if retries are happening if the failed request takes longer. We rejected this option because the results could vary depending on the load of system resources or other factors. But please let us know if you recommend a testing approach that would work.

## Does this PR include user facing changes?

- [x] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Notes
- Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
- Do not hesitate to use `@vectordotdev/vector` to reach out to us regarding this PR.
- The CI checks run only after we manually approve them.
  - We recommend adding a `pre-push` hook, please see [this template](https://github.com/vectordotdev/vector/blob/master/CONTRIBUTING.md#Pre-push).
  - Alternatively, we recommend running the following locally before pushing to the remote branch:
    - `cargo fmt --all`
    - `cargo clippy --workspace --all-targets -- -D warnings`
    - `cargo nextest run --workspace` (alternatively, you can run `cargo test --all`)
    - `./scripts/check_changelog_fragments.sh`
- After a review is requested, please avoid force pushes to help us review incrementally.
  - Feel free to push as many commits as you want. They will be squashed into one before merging.
  - For example, you can run `git merge origin master` and `git push`.
- If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
  run `cargo vdev build licenses` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).


<!-- Please list any issues closed by this PR. -->


- Closes: #10870


<!-- Any other issues or PRs relevant to this PR? Feel free to list them here. -->

thomasqueirozb commented on Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "vector"
-version = "0.48.0"
```
Please don't bump the vector version, we have our own release process

jchap-pnnl replied:
Hmm, I'm not sure what to do about this. I bumped up the version because without it, running `cargo vdev check version` gives me this error:
![image](https://github.com/user-attachments/assets/55eb29f5-3d5a-4f5b-90b1-1de36b7ed3ea)
Is there another way to resolve that error without changing the vector version locally?

jchap-pnnl replied:
I've reverted the change.

thomasqueirozb commented on src/sinks/s3_common/config.rs:
```diff
@@ -312,15 +312,73 @@ impl From<S3CannedAcl> for ObjectCannedAcl {
+
+    /// Retry specific errors.
+    ///
+    /// A vector list of status codes matching specific error types that trigger failed
+    /// service retry attempts. The list is ignored if `retry_all_errors` is true.
```
```suggestion
    /// A list of HTTP status codes matching specific error types that trigger failed
    /// service retry attempts. This list is ignored if `retry_all_errors` is true.
```

jchap-pnnl commented on website/cue/reference/components/sinks/base/aws_s3.cue:
```diff
@@ -1044,6 +1044,36 @@ base: components: sinks: aws_s3: configuration: {
+				description: """
+					Retry specific errors.
+
+					A vector list of status codes matching specific error types that will trigger
```
I've made this change.

pront commented on src/sinks/s3_common/config.rs:
```diff
@@ -312,15 +312,73 @@ impl From<S3CannedAcl> for ObjectCannedAcl {
     }
 }
 
+fn check_response(res: &HttpResponse, errors_to_retry: Option<Vec<u16>>) -> bool {
```
```suggestion
fn is_retryable_response(res: &HttpResponse, errors_to_retry: Option<Vec<u16>>) -> bool {
```

jchap-pnnl replied:
Good suggestion, but I'm going to change the function name to `is_retriable_response`, since I see 57 spots in the codebase where "retryable" is spelled "retriable." I'm not sure which is the correct way to spell it, but will side with consistency for now.

pront commented on src/sinks/s3_common/config.rs:
```diff
@@ -312,15 +312,73 @@ impl From<S3CannedAcl> for ObjectCannedAcl {
+    }
+}
+
+fn configured_to_retry(
```
```suggestion
fn should_retry_error(
```

pront commented on src/sinks/s3_common/config.rs:
```diff
@@ -312,15 +312,73 @@ impl From<S3CannedAcl> for ObjectCannedAcl {
+#[configurable_component]
 #[derive(Debug, Clone)]
-pub struct S3RetryLogic;
+pub struct S3RetryLogic {
+    /// Retry all errors.
+    ///
+    /// The sink retries all failed service calls.
+    pub retry_all_errors: Option<bool>,
+
+    /// Retry specific errors.
+    ///
+    /// A list of HTTP status codes matching specific error types that trigger failed
+    /// service retry attempts. This list is ignored if `retry_all_errors` is true.
+    pub errors_to_retry: Option<Vec<u16>>,
+}
```
```Rust
#[derive(Debug, Clone, Default)]
pub enum RetryStrategy {
    /// Don’t retry any errors
    #[default]
    None,

    /// Retry on *all* errors
    All,

    /// Retry only on these HTTP status codes
    StatusCodes(Vec<u16>),
}
```

Something like this seems a more natural

jchap-pnnl replied:
I've made these changes. Thanks for the suggestions.

pront reviewed (commented):
Left some nits

pront commented:
We will review again once the merge conflicts are resolved.

thomasqueirozb commented:
Just ran `git fetch && git checkout origin/master -- Cargo.toml Cargo.lock LICENSE-3rdparty.csv && git merge origin/master` to fix your merge conflicts. Cargo.lock still shows changes which is weird but you should be good now to implement the suggestions made in this PR 👍

jchap-pnnl commented:
Awesome, thanks for your help, @thomasqueirozb!

pront commented:
I will take a look @jchap-pnnl

jchap-pnnl commented:
Great, thanks!

jchap-pnnl commented:
I reran the tests, verified the functionality still works, and updated the PR description. I've addressed everything I'm aware of. Please resume reviewing. Thanks!
